src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyCallQueue.cs

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Hosting
+{
+    // A mechanism for queuing JS-to-.NET calls so they aren't nested on the call stack and hence
+    // have the same ordering behaviors as in Blazor Server. This eliminates serveral inconsistency
+    // problems and bugs that otherwise require special-case solutions in other parts of the code.
+    //
+    // The reason for not using an actual SynchronizationContext for this is that, historically,
+    // Blazor WebAssembly has not enforced any rule around having to dispatch to a sync context.
+    // Adding such a rule now would be too breaking, given how component libraries may be reliant
+    // on being able to render at any time without InvokeAsync. If we add true multithreading in the
+    // future, we should start enforcing dispatch if (and only if) multithreading is enabled.
+    // For now, this minimal work queue is an internal detail of how the framework dispatches
+    // incoming JS->.NET calls and makes sure they get deferred if a renderbatch is in process.
+
+    internal static class WebAssemblyCallQueue
```

Comment:
This looks very similar to the RendererSynchronizationContext is there a way we could "reuse" that here instead of having a separate work queue?

These things are tricky to test implement and can lead to subtle bugs that are hard to track down

Reply:
The main issue with using a sync context is the back-compat about enforcement of being "on" the sync context (discussed above), but even besides that, the existence of synchronous interop makes WebAssembly a different kind of runtime environment for which we can't assume that `RendererSynchronizationContext` has the right behaviors. For example, incoming calls may already be on the sync context but we still need to defer their work. It's not the right fit for this runtime.

I know your concern is more future-thinking and looks towards a scenario where we have real multithreading. My expectation is that when/if we get to that, we'll be in a position to recreate the same threading semantics as Server pretty much exactly (assuming we want to), but that's not the case today and shouldn't be considered in scope for fixing this bug. My goal in this PR is to wipe out the largest class of possible inconsistency issues I can while not regressing compat or perf.